compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/calls/Arguments.kt

```diff
@@ -239,25 +236,59 @@ fun Candidate.resolvePlainArgumentType(
             capturedType
 
     // If the argument is of functional type and the expected type is a suspend function type, we need to do "suspend conversion."
+    if (expectedType != null) {
+        argumentTypeWithSuspendConversion(
+            session, context.bodyResolveComponents.scopeSession, expectedType, argumentTypeForApplicabilityCheck
+        )?.let {
+            argumentTypeForApplicabilityCheck = it
+            substitutor.substituteOrSelf(argumentTypeForApplicabilityCheck)
+            usesSuspendConversion = true
+        }
+    }
+
+    checkApplicabilityForArgumentType(
+        csBuilder, argumentTypeForApplicabilityCheck, expectedType, position, isReceiver, isDispatch, sink, context
+    )
+}
+
+private fun argumentTypeWithSuspendConversion(
+    session: FirSession,
+    scopeSession: ScopeSession,
+    expectedType: ConeKotlinType,
+    argumentType: ConeKotlinType
+): ConeKotlinType? {
     // TODO: should refer to LanguageVersionSettings.SuspendConversion
-    if (expectedType?.isSuspendFunctionType(session) == true &&
-        argumentTypeForApplicabilityCheck.isBuiltinFunctionalType(session) &&
-        !argumentTypeForApplicabilityCheck.isSuspendFunctionType(session)
+    // Expect the expected type to be a suspend functional type.
+    if (!expectedType.isSuspendFunctionType(session)) {
+        return null
+    }
+
+    // Either the argument type is a non-suspend functional type
+    if (argumentType.isBuiltinFunctionalType(session) && !argumentType.isSuspendFunctionType(session)) {
```

Comment:
The same logic about unification may be applied here as well

Reply:
I tried this place as well, but no luck. In particular, for the case of direct functional type, building the functional type through an invoke symbol bothers resolving valid suspend conversion cases with generic types. Maybe this is a hint of finding invoke symbol is somewhat wrong? Anyhow, I ended up splitting them again, and this time, for proper type constraint generation, even not using invoke symbol for functional type.

Reply:
Maybe my previous comments may help here as well
If not, please do not hesitate to ask again

Reply:
Finding base invoke from unsubstituted scope and finding contributed invoke from the scope of argument type help me figure out the difference between two. Here, as constraints, we need type substituted one, and I finally managed to unify the logic.